### DISCUSSION TITLE
Popup Size Restriction

### DISCUSSION THREAD

**Discussion by QyQj** [dotnet/wpf#9830] [Q&A] [answered]
I was trying to pop up some dialog just a little bit smaller than the screen instead of opening a new window

But it seems to be impossible, the `Height` of `DesiredSize` of popup was always set to 1200 (my screnn has a resolution of 2560X1600)  

I spent a lot of time to debug (it was so anoying) and finally find these codes in popup.cs

```     
// Limit size to 75% of maxDimension's area and restrict to be smaller than limitDimension
internal Size RestrictSize(Size desiredSize)
{
      // Make sure screen bounds and limit dimensions are up to date
      Rect targetBounds, screenBounds;
      Size limitSize;
      GetPopupRootLimits(out targetBounds, out screenBounds, out limitSize);

     // Convert from popup's space to screen space
     desiredSize = (Size)_secHelper.GetTransformToDevice().Transform((Point)desiredSize);

     desiredSize.Width = Math.Min(desiredSize.Width, screenBounds.Width);
     desiredSize.Width = Math.Min(desiredSize.Width, limitSize.Width);

     double maxHeight = RestrictPercentage * screenBounds.Width * screenBounds.Height / desiredSize.Width;

     desiredSize.Height = Math.Min(desiredSize.Height, screenBounds.Height);
     desiredSize.Height = Math.Min(desiredSize.Height, maxHeight);
     desiredSize.Height = Math.Min(desiredSize.Height, limitSize.Height);

     // Convert back from screen space to popup's space
    desiredSize = (Size)_secHelper.GetTransformFromDevice().Transform((Point)desiredSize);

    return desiredSize;
}
```

Why does the popup's size is always limited to smaller than 75% of available area and the percentage is not public  and how can I change it?

**Comment by lindexi** [comment]
I read the code of wpf, but I can not find the answer.

https://github.com/dotnet/wpf/blob/4fe1516393e59be84249dacfb275c1e19068eabe/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Primitives/Popup.cs#L2619


As https://github.com/dotnet/wpf/blob/4fe1516393e59be84249dacfb275c1e19068eabe/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Primitives/Popup.cs#L2222 says, 

> This prevents the popup from overlapping the placement target.

**Comment by miloush** [comment, marked as answer]
This is for security reasons, creating a full-screen or larger popup could block users form interacting with the app. You cannot change it, it's a constant. If you want to show a dialog, can't you use `Window.ShowDialog()`?

**Reply by QyQj** [reply]
Thank you for your answer

Exactly I was using the `DialogHost` of  MaterialDesignWpf to popup a Dialog instead of using a window

I do understand this security design, but in my case, even without the limit of 75% ,there are still enough security space and user will not be blocked (The size of popup content was smaller than the screen), so I think the limit was so rigid

And I think the limit should be set both on `Width` and `Height` ......

Any way, I have changed the  size with visualtree

**Reply by miloush** [reply]
I guess the average display resolutions are nowadays slightly different than 20 years ago... what popup size were you trying to show, what is the use case?

**Reply by QyQj** [reply]
I have a management system which will be used during user using the MainWindow（The MainWindow is set to fullscreen and has a resolution of 1920X1080 or higher）

I set it as the popup content which has a size as same as the MainWindow, also I set a margin to avoid the popup cover the MainWindow

Exactly if I didn't set the margin it will be ok too, there is a button to close the popup ,the margin is just for some visual effect (there will be a dark transparent mask behind the popup content)

And after all, use a new window to show dialog is more convenient but we're using the MaterialDesignTheme, to keep the unification in design I decide to use the `DialogHost` in MaterialDesignTheme which use popup to show dialog

**Reply by RandyPJ** [reply]
I think it is convenient to make a styled  dialog window(a big one) and make use of the material design and add shadows to it so it looks cool.

The advantage of making your own dialog is that is independent and that you can pass whatever content you need based on a condition, lets say a viewmodel.
